crates/uv-client/src/base_client.rs

```diff
@@ -440,6 +482,59 @@ impl BaseClient {
     }
 }
 
+/// Wrapper around [`ClientWithMiddleware`] that manages redirects.
+#[derive(Debug, Clone)]
+pub struct RedirectClientWithMiddleware<'a>(&'a ClientWithMiddleware);
+
+impl RedirectClientWithMiddleware<'_> {
+    /// Executes a request. If the response is a 302 redirect, executes the
+    /// request again with the redirect location URL (up to a maximum number
+    /// of redirects).
+    ///
+    /// Unlike the built-in reqwest redirect policies, this sends the
+    /// redirect request through the entire middleware pipeline again.
+    pub async fn execute(&self, req: Request) -> reqwest_middleware::Result<Response> {
+        let mut request = req;
+        let max_redirects = 10;
+
+        for redirects in 0..=max_redirects {
+            let result = self
+                .0
+                .execute(request.try_clone().expect("HTTP request must be cloneable"))
+                .await;
+            if result.is_err() || redirects == max_redirects {
+                return result;
+            }
+            let response = result.unwrap();
+
+            // Handle redirect if we receive a 302
+            if response.status() == StatusCode::FOUND {
```

Comment:
(*) Should this apply to other 30x redirect codes too, or only 302?

Reply:
I was trying to limit this to the proxy-redirecting case that motivated the change (in case there are other edge cases we'd have to handle). But I suppose it probably can't hurt to re-run the middleware pipeline on any redirect. Would that be your preference?

Reply:
I'd opportunistically handle all applicable redirect codes, it feels more coherent and helps users that encounter other redirect codes in the future.